test/metabase/query_processor_test/date_time_zone_functions_test.clj

```diff
@@ -169,3 +177,141 @@
                  :fields [[:field (mt/id :times :index) nil]]}]]]
         (testing title
           (is (= expected (test-date-extract query))))))))
+
+(defn- date-math
+  [op x amount unit col-type]
+  (let [amount (if (= op :date-add)
+                 amount
+                 (- amount))
+        fmt    (cond
+                 ;; the :date column of :presto should have this format too,
+                 ;; but the test data we created for presto is datetime even if we define it as date
+                 (and (= driver/*driver* :presto) (#{:text-as-date} col-type))
+                 "yyyy-MM-dd"
+
+                 :else
+                 "yyyy-MM-dd HH:mm:ss")]
+    (t/format fmt (u.date/add x unit amount))))
+
+(defn- mongo-major-version [db]
+  (-> (get-in db [:details :version])
+      (str/split #"\.")
+      first
+      parse-long))
+
+(deftest date-math-tests
+  (mt/dataset times-mixed
+    (mt/test-drivers (disj (mt/normal-drivers-with-feature :date-arithmetics) :mongo)
+      (testing "date arithmetic with datetime columns"
+        (doseq [[col-type field-id] [[:datetime (mt/id :times :dt)] [:text-as-datetime (mt/id :times :as_dt)]]
+                op                  [:date-add :date-subtract]
+                unit                [:year :quarter :month :day :hour :minute :second]
+                [expected query]
+                [[[[(date-math op #t "2004-03-19 09:19:09" 2 unit col-type)] [(date-math op #t "2008-06-20 10:20:10" 2 unit col-type)]
+                   [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)] [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)]]
+                  {:expressions {"expr" [op [:field field-id nil] 2 unit]}
+                   :fields      [[:expression "expr"]]}]
```

Comment:
I know tuples like `[expected query]` are more convenient than maps but seeing `[[[[` makes my eyes bleed... I think 

```clj
{:expected ...
 :query ...}
```

would be a bit more readable. Maybe if you move this stuff into a top-level `def` you can do that instead.